pkg/middlewares/retry/retry.go

```diff
@@ -72,34 +82,61 @@ func (r *retry) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	}
 
 	attempts := 1
+	backOff := r.newBackOff()
+	currentInterval := 0 * time.Millisecond
 	for {
-		shouldRetry := attempts < r.attempts
-		retryResponseWriter := newResponseWriter(rw, shouldRetry)
-
-		// Disable retries when the backend already received request data
-		trace := &httptrace.ClientTrace{
-			WroteHeaders: func() {
-				retryResponseWriter.DisableRetries()
-			},
-			WroteRequest: func(httptrace.WroteRequestInfo) {
-				retryResponseWriter.DisableRetries()
-			},
-		}
-		newCtx := httptrace.WithClientTrace(req.Context(), trace)
+		select {
+		case <-time.After(currentInterval):
 
-		r.next.ServeHTTP(retryResponseWriter, req.WithContext(newCtx))
+			shouldRetry := attempts < r.attempts
+			retryResponseWriter := newResponseWriter(rw, shouldRetry)
 
-		if !retryResponseWriter.ShouldRetry() {
-			break
-		}
+			// Disable retries when the backend already received request data
+			trace := &httptrace.ClientTrace{
+				WroteHeaders: func() {
+					retryResponseWriter.DisableRetries()
+				},
+				WroteRequest: func(httptrace.WroteRequestInfo) {
+					retryResponseWriter.DisableRetries()
+				},
+			}
+			newCtx := httptrace.WithClientTrace(req.Context(), trace)
+
+			r.next.ServeHTTP(retryResponseWriter, req.WithContext(newCtx))
+
+			if !retryResponseWriter.ShouldRetry() {
+				return
+			}
+
+			currentInterval = backOff.NextBackOff()
+
+			attempts++
 
-		attempts++
+			log.FromContext(middlewares.GetLoggerCtx(req.Context(), r.name, typeName)).
+				Debugf("New attempt %d for request: %v", attempts, req.URL)
 
-		log.FromContext(middlewares.GetLoggerCtx(req.Context(), r.name, typeName)).
-			Debugf("New attempt %d for request: %v", attempts, req.URL)
+			r.listener.Retried(req, attempts)
 
-		r.listener.Retried(req, attempts)
+		case <-req.Context().Done():
+			return
+		}
+	}
+}
+
+func (r *retry) newBackOff() nexter {
+	if r.attempts < 2 || r.initialInterval <= 0 {
+		return &backoff.ZeroBackOff{}
 	}
+
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = r.initialInterval
+
+	// calculate the multiplier that will set MaxInterval = 2*InitialInterval
```

Comment:
Thanks for your previous changes, i think the comment should clarify the purpose here, and not confuse with the max interval, what do you think ?
```suggestion
	// calculate the multiplier for the given number of attempts
	// so that applying the multiplier for the given number of attempts will not exceed 2 times the initial interval
	// it allows to control the progression along the attempts
```